### ISSUE TITLE
HiddenField check does not work well with builders

### ISSUE THREAD

**Issue by mkordas** [checkstyle/checkstyle#619]
[HiddenField](http://checkstyle.sourceforge.net/config_coding.html#HiddenField) check could be extended to handle builder methods better. Right now it supports them if `setterCanReturnItsClass` is set to `true` and method name starts from `set`. 

Builder method prefix should be more configurable. E.g. very common pattern is to start builder methods with prefix `with`:

```java
class Builder {
    private int number;

    public Builder withNumber(int number) {
        this.number = number; // violation
    }

    public Builder setNumber(int number) {
        this.number = number; // works fine with enabled setterCanReturnItsClass
    }
}
```

**Comment by romani** [comment]
hmmmm, that we planned to do that, but for some reason it was lost in update, all discussion is here - https://groups.google.com/forum/#!topic/checkstyle-devel/aPNYF3RC_yU

**Comment by rdiachenko** [comment]
There are builders which don't use prefixes in methods which can be a tricky case:
```
class Builder {
  private int id;
  private String name;
  
  public Builder id(int id) {
    this.id = id;
	return this;
  }
  
  public Builder name(String name) {
    this.name = name;
	return this;
  }
}
```

What if we remove ```setterCanReturnItsClass``` option at all? It complicates the things and leads to a lot of false positives. Providing a new option for prefixes doesn't seem good to me as there seemed to be other tricky cases as the aforementioned one.

If we look in general, I think most of such setters ```Foo setProp(int prop) { this.prop = prop; }``` which return the including class can be found in Builders. But we can ignore Builders at all as they all have similar structure.

**Comment by priimak** [comment]
Well. I somewhat agree with idea of getting rid of setterCanReturnItsClass option, even though I am the
one who proposed it in the first place. The ultimate question is how to identify methods in which hidden
field check is to be ignored. At the moment we look at the following elements of method signature: 
1) What is returned (void or a class in which method is a member of) 
2) Method name and single parameter variable should be in the following form

    set$Name(type $name), where $Name == capitalize($name)

Some of my coworkers suggested simple regexp for method name to identify such methods.
For example $methodName =~ /^set.*/ but that proposal was rejected. Any other ideas?

**Comment by mkordas** [comment]
My idea is to ignore HiddenField check in all methods which:
* return containing class
* accept one parameter
* method name ends with capitalized parameter name OR method has the same name as parameter

This would ignore violations on all kinds of builders, e.g.:
```java
class Builder {
    private int number;

    public Builder number(int number) {
        this.number = number;
    }

    public Builder withNumber(int number) {
        this.number = number;
    }

    public Builder setNumber(int number) {
        this.number = number;
    }

    public Builder usingNumber(int number) {
        this.number = number;
    }

    public Builder havingNumber(int number) {
        this.number = number;
    }
}
```

**Comment by romani** [comment]
thanks to all for bringing that problem back to attention, and sharing their experience and ideas.

Lets come back to origin of that Check – it purpose to forbid using local variables names like filed and avoid mistakes like - https://github.com/checkstyle/checkstyle/commit/83e23be070d55f7a408161876cc310fdd5ff113e.
The first problem is completely valid and no discussion is required. But the second is very controversial as it conflicting with idea of reusing good names from signature of methods  as suggested names in caller code (during code code generation in IDE). 
It could be resolved by "require this" validation inside methods on conflicting names
But changing the name of parameter is long-lasting pain to somebody who use that method, especially if that is frequently used method.


It will be just waste of time and spread ugly names population code. 
Remember that in 21th century engineers do not have time to read javadoc for all method they use.
Developers rely name on method and names of parameters and understand how it works from signature.

Lets image class that :
```
class Test { 

private int timeLimit;
private Function fuction;

sort(col, fun, tlimit) { …. }

}
```
that will suggest to user of this code to have :
```
Collection  col = ….;
Function fun = ….;
int tlimit = …;
testObject.sort(col, fun, tlimit);
```
instead of 
```
Collection  collection = ….;
Function function = ….;
int timeLimit = …;

testObject.sort(column, function, timelimit);
```

So by resolving internal problem of class Test we damage code of all users (all code that use that method), it is not right approach!!!!  Problems of inner class design should not go outside of that class.

One more point, but very vital point – WHY THAT CHECK TREAT SETTERS AS SOMETHING SPECIAL ? 
In example above your can see that this is problem in all methods. 
The whole idea of adding Builder  class as smth special come from desire to name good suggested names in caller code (code that call a method).
So we should stop thinking about setters as smth special !!!!!! 

It was a mistake a lot time ago to do that option – ignoreSetter.  As user wanted to skip violation for some names , so to option to skip violation by methods names had to be provided.

Ok ... , we can not remove that option easily as it will braking compatibility  and still see a workaround there.

So my suggestion is to:
- remove PARAMETER_DEF from default of token in Check to not do force user generate ugly names for variables.
- create another option to   skip violation for some names , so to option to skip violation by methods names should to be provided



2 @rdiachenko – if such style take a place, user should not use PARAMETER_DEF as target token. And should use “RequireThis” Check (https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java).

2 @mkordas, please see reply to @rdiachenko .


For new option: 
Using simple set of  prefixes will not work as we will come to conflicts with words that have "set" 
and "with" - "settlings", "settlements" , "withdrawal", "withhold"......, you can use http://www.wordplays.com/words-that-start-with
to estimate false-positives.
We have to use RegExp like - "^(set|with)[A-Z].*", but default should be  "^(set)[A-Z].*"


Newly added option allow oldstyle/convention named setters to be recognized. So for now we support only 
conventional setters (we do not support setters named in functional style), it is difficult to recognize.
So new option make sense , and will not be removed.
The problem in that option is that it show that next step is also possible to implement.
Option should stay as it make Check more smart for conventional Java developers.


JavaDoc/Xdoc should be extended to warn user on usage of PARAMETER_DEF token, examples should be provided to user.

**Comment by andrewgaul** [comment]
Checkstyle should optionally allow Google Auto-style Builders which have no prefix:

https://github.com/google/auto/tree/master/value#builders

Even though Checkstyle will not see Google Auto Builders due to code generation, several applications use this style directly in source.

**Comment by romani** [comment]
@andrewgaul , thanks a lot , this is just another proove that such Check should not validate method arguments by default , as there is no way to detect properly Builders (it is very offen case when users modify patterns for their needs).

**Comment by vanniktech** [comment]
If Checkstyle can't detect all of the Builders, why don't we go with the `set` approach? Where one can pass prefixes like `set,with,...`

While this is not an optimal solution we could go and advance and maybe iterate in the future again over this problem.

**Comment by mkordas** [comment]
@vanniktech +1 for the iterative approach idea

**Comment by romani** [comment]
I am ok with starting implementation of this issue only in following conditions:

1) new option will be RegExp see a reason at the end of https://github.com/checkstyle/checkstyle/issues/619#issuecomment-74564859 post.

2) 
> JavaDoc/Xdoc should be extended to warn user on usage of PARAMETER_DEF token, examples should be provided to user.

that should be done too to address Builders with design as https://github.com/checkstyle/checkstyle/issues/619#issuecomment-75291840

3) Javadoc/xdoc of check should be updated to explain user that Check is not ideal and could not reconcile all design/naming approaches. So PARAMETER_DEF should be removed if user do not want to have extra violations. future implementation RequireThis Check should be used to avoid naming problems see https://github.com/checkstyle/checkstyle/issues/2362 for more details.

@mkordas , @vanniktech , @andrewgaul , @priimak are you agree ?

**Comment by vanniktech** [comment]
Sounds good

**Comment by mkordas** [comment]
@romani +1 from me

**Comment by priimak** [comment]
Sound good to me.

**Comment by romani** [comment]
@andrewgaul , please share your opinion. (if no reply in 7 days, issue become "approved" for implementation)

**Comment by vanniktech** [comment]
@romani it's time for marking this as approved. Also has there been any progress on this one?

**Comment by TWiStErRob** [comment]
@romani bump, really need for that regex option.

**Comment by eirirlar** [comment]
Bump

**Comment by romani** [comment]
ok, issue is approved. ALL are welcome with PRs.

**Comment by imrel** [comment]
Hi @romani and others, 

1. "new option will be RegExp" - I dont see it helping on cases where builder method has no prefix
2. Is there a way to disable for classes by pattern i.e "*Builder"
3. same as 2 but for inner classes

Thanks.

**Comment by romani** [comment]
Yes, try our xpath SuppressionFilter https://checkstyle.org/config_filters.html#SuppressionXpathFilter, in next release it will be possible to use suppression without separate file.

**Comment by DakshRJain737** [comment]
@romani Is this issue still open? If yes, then
From the above conversation, I have come to a conclusion to include two features:

1. In builder classes, a method can have a prefix as `with`.
2. In builder classes, a method can have same name as the class field.
Only methods which return their own class are to be ignored by the HiddenFieldCheck.
```
public class Demo {
    private Long id;

    public Demo withId(long id) {
        this.id = id;
        return this;
    }

    public Demo id(long id) {
        this.id = id;
        return this;
    }
}
```
If there is any thing else that I might have missed, then correct me please.

**Comment by DakshRJain737** [comment]
https://checkstyle.sourceforge.io/checks/coding/hiddenfield.html 
The Example1 of the HiddenFieldCheck is stated below:
```
class Example1 {

  private String field;
  private String testField;

  Example1(String testField) { // violation, 'field' hides a field
  }
  void method(String param) {
    String field = param; // violation, ''field' hides a field'
  }
  void setTestField(String testField) { // violation, ''testField' hides a field'
    this.field = field;
  }
  Example1 setField(String field) { // violation, ''field' hides a field'
    this.field = field;
    return null;
  }
  abstract class Inner {
    abstract int method(String field); // violation, ''field' hides a field'
  }
}
```

Am I the only one who is finding inconsistency in this example?

The constructor violation message,  `setTestField` body is wrong.
